### PR TITLE
Verify SharedGroup in handleAsyncTransactionCompleted

### PR DESCRIPTION
**Pull request by beeender**
In a rare case, if handleAsyncTransactionCompleted posted before set
handler to null, the handleAsyncTransactionCompleted will be called with
a closed Realm.

It is very difficult to construct a test case for this.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 * Throw a proper exception when operating on a non-existing field with the dynamic API (#3292).
 * `DynamicRealmObject.setList` should only accept `RealmList<DynamicRealmObject>` (#3280).
+* Fixed a concurrency crash which might happen when `Realm.executeTransactionAsync()` called with a `OnSuccess` and Realm closed before `OnSuccess` called.
```
Comment (cmelchior) [inline CHANGELOG.md]: Fixed a concurrency crash which might happen when `Realm.executeTransactionAsync()` tried to cal `onSucess` after the Realm was closed.

File: realm/realm-library/src/main/java/io/realm/HandlerController.java [hunk 2]
```diff
@@ -156,10 +156,13 @@ public boolean handleMessage(Message message) {
      * @param onSuccess onSuccess callback to run for the async transaction that completed.
      */
     public void handleAsyncTransactionCompleted(Runnable onSuccess) {
-        if (onSuccess != null) {
-            pendingOnSuccessAsyncTransactionCallbacks.add(onSuccess);
+        // Same reason with handleMessage
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/HandlerController.java]: Same reason as handeMessage()

**Comment by cmelchior** [comment]
Nice catch 👍

**Comment by nhachicha** [comment]
👍

**Pull request merged**